Data/HashMap/Internal.hs

```diff
@@ -1771,37 +1756,156 @@ differenceWith f a b = foldlWithKey' go empty a
+      | b1 .&. m2 == 0 = Empty
+      | otherwise = go (s + bitsPerSubkey) (A.index ary1 i) t2
+      where
+        m2 = mask h2 s
+        i = sparseIndex b1 m2
+    go s t1@(Collision h1 _ls1) (BitmapIndexed b2 ary2)
+      | b2 .&. m1 == 0 = Empty
+      | otherwise = go (s + bitsPerSubkey) t1 (A.index ary2 i)
+      where
+        m1 = mask h1 s
+        i = sparseIndex b2 m1
+    go s (Full ary1) t2@(Collision h2 _ls2) = go (s + bitsPerSubkey) (A.index ary1 i) t2
+      where
+        i = index h2 s
+    go s t1@(Collision h1 _ls1) (Full ary2) = go (s + bitsPerSubkey) t1 (A.index ary2 i)
+      where
+        i = index h1 s
+
+    intersectionArray s b1 b2 ary1 ary2
+      -- don't create an array of size zero in intersectionArrayBy
+      | b1 .&. b2 == 0 = Empty
+      | otherwise = runST $ do
+        (b, len, ary) <- intersectionArrayBy (go (s + bitsPerSubkey)) b1 b2 ary1 ary2
+        case len of
+          0 -> pure Empty
+          1 -> A.read ary 0
+          _ -> bitmapIndexedOrFull b <$> (A.unsafeFreeze =<< A.shrink ary len)
+{-# INLINE intersectionWithKey# #-}
+
+intersectionArrayBy ::
+  ( HashMap k v1 ->
+    HashMap k v2 ->
+    HashMap k v3
+  ) ->
+  Bitmap ->
+  Bitmap ->
+  A.Array (HashMap k v1) ->
+  A.Array (HashMap k v2) ->
+  ST s (Bitmap, Int, A.MArray s (HashMap k v3))
+intersectionArrayBy f !b1 !b2 !ary1 !ary2 = do
+  mary <- A.new_ $ popCount bIntersect
+  -- iterate over nonzero bits of b1 .&. b2
+  let go !i !i1 !i2 !b !bFinal
+        | b == 0 = pure (i, bFinal)
+        | testBit $ b1 .&. b2 = do
+          x1 <- A.indexM ary1 i1
+          x2 <- A.indexM ary2 i2
+          case f x1 x2 of
+            Empty -> go i (i1 + 1) (i2 + 1) b' (bFinal .&. complement m)
+            _ -> do
+              A.write mary i $! f x1 x2
+              go (i + 1) (i1 + 1) (i2 + 1) b' bFinal
+        | testBit b1 = go i (i1 + 1) i2 b' bFinal
+        | otherwise = go i i1 (i2 + 1) b' bFinal
+        where
+          m = 1 `unsafeShiftL` countTrailingZeros b
+          testBit x = x .&. m /= 0
+          b' = b .&. complement m
+  (maryLen, bFinal) <- go 0 0 0 bCombined bIntersect
+  pure (bFinal, maryLen, mary)
+  where
+    bCombined = b1 .|. b2
+    bIntersect = b1 .&. b2
+{-# INLINE intersectionArrayBy #-}
+
+intersectionCollisions :: Eq k => (k -> v1 -> v2 -> (# v3 #)) -> A.Array (Leaf k v1) -> A.Array (Leaf k v2) -> ST s (Int, A.MArray s (Leaf k v3))
+intersectionCollisions f ary1 ary2 = do
+  mary2 <- A.thaw ary2 0 $ A.length ary2
+  mary <- A.new_ $ A.length ary1 + A.length ary2
+  let go i j
+        | i >= A.length ary1 || j >= A.lengthM mary2 = pure j
+        | otherwise = do
+          L k1 v1 <- A.indexM ary1 i
+          searchSwap k1 j mary2 >>= \case
+            Just (L _k2 v2) -> do
+              let !(# v3 #) = f k1 v1 v2
+              A.write mary j $ L k1 v3
+              go (i + 1) (j + 1)
+            Nothing -> do
+              go (i + 1) j
+  maryLen <- go 0 0
+  pure (maryLen, mary)
```

Comment:
Should we go ahead and produce the actual `HashMap` fragment here?

Reply:
> Also if we pass the function around through recursion, then we wouldn't be able to implement `intersection` in terms of `intersectionWithKey` right?

Right. There are dizzyingly many options with equally many trade-offs. I hate it. That said, I wonder if we should be implementing `intersection` separately anyway, since it can get sharing in the leaves. Ugh.

Reply:
> Should we go ahead and produce the actual `HashMap` fragment here?

Yeah, I think that would be clearer.